tests/shared/src/test/scala/cats/effect/CallbackStackSpec.scala

```diff
@@ -21,10 +21,40 @@ class CallbackStackSpec extends BaseSpec {
   "CallbackStack" should {
     "correctly report the number removed" in {
       val stack = CallbackStack[Unit](null)
-      val pushed = stack.push(_ => ())
-      val handle = pushed.currentHandle()
-      pushed.clearCurrent(handle)
-      stack.pack(1) must beEqualTo(1)
+      val handle = stack.push(_ => ())
+      stack.push(_ => ())
+      val removed = stack.clearHandle(handle)
+      if (removed)
+        stack.pack(1) must beEqualTo(0)
+      else
+        stack.pack(1) must beEqualTo(1)
+    }
+
+    "handle race conditions in pack" in real {
+
+      IO(CallbackStack[Unit](null)).flatMap { stack =>
+        val pushClearPack = for {
+          handle <- IO(stack.push(_ => ()))
+          removed <- IO(stack.clearHandle(handle))
+          packed <- IO(stack.pack(1))
+        } yield (if (removed) 1 else 0) + packed
+
+        pushClearPack
+          .both(pushClearPack)
+          .product(IO(stack.pack(1)))
+          .flatMap { case ((x, y), z) => IO((x + y + z) must beEqualTo(2)) }
+          .replicateA_(1000)
+          .as(ok)
+      }
+    }
+
+    "pack runs concurrently with clear" in real {
+      IO {
+        val stack = CallbackStack[Unit](null)
+        val handle = stack.push(_ => ())
+        stack.clearHandle(handle)
+        stack
+      }.flatMap(stack => IO(stack.pack(1)).both(IO(stack.clear()))).replicateA_(1000).as(ok)
```

Comment:
Same comment as above regarding parallelism factor.

```suggestion
      }.flatMap(stack => IO(stack.pack(1)).both(IO(stack.clear()))).parReplicateA_(1000).as(ok)
```